### PR TITLE
feat(IdGenerator): Add IdGenerator.SimpleGuid and IdGenerator.SequentialGuid and IdGenerator.Snowflake

### PR DESCRIPTION
**Pull request by zhenlei520**

### CODE REVIEW

File: Masa.Contrib.sln [hunk 1]
```diff
@@ -211,10 +211,28 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Masa.BuildingBlocks.Service
...
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "IdGenerator", "IdGenerator", "{E6363F59-2BA4-4AA7-8578-C433A2C2567F}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Masa.Contrib.Data.IdGenerator.Snowflake.BenchmarkDotnet.Tests", "test\Masa.Contrib.Data.IdGenerator.Snowflake.BenchmarkDotnet.Tests\Masa.Contrib.Data.IdGenerator.Snowflake.BenchmarkDotnet.Tests.csproj", "{D8E85337-A779-48FC-B822-CE8A850661DC}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Masa.BuildingBlocks.Data.MappingExtensions", "src\BuildingBlocks\MASA.BuildingBlocks\src\Data\Masa.BuildingBlocks.Data.MappingExtensions\Masa.BuildingBlocks.Data.MappingExtensions.csproj", "{30DC35DF-9D86-443C-B26C-9053E6FCA434}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Masa.Contrib.Data.IdGenerator.SimpleGuid", "src\Data\IdGenerator\Masa.Contrib.Data.IdGenerator.SimpleGuid\Masa.Contrib.Data.IdGenerator.SimpleGuid.csproj", "{24E5BF57-33B6-4D5B-A32F-2F64D1A5954F}"
```
Comment (doddgu) [inline Masa.Contrib.sln:232]: simple -> normal

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.SequentialGuid/SequentialGuidGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Contrib.Data.IdGenerator.SequentialGuid;
+
+/// <summary>
+/// This code is taken from jhtodd/SequentialGuid https://github.com/jhtodd/SequentialGuid/blob/master/SequentialGuid/Classes/SequentialGuid.cs
```
Comment (doddgu) [inline src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.SequentialGuid/SequentialGuidGenerator.cs:7]: If there is no license, the copyright is available by default

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.SimpleGuid/README.zh-CN.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
+﻿中 | [EN](README.md)
+
+## Masa.Contrib.Data.IdGenerator.SimpleGuid
+
+Masa.Contrib.Data.IdGenerator.SimpleGuid是一个简单的id构造器，提供Guid类型的唯一标识
```
Comment (doddgu) [inline src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.SimpleGuid/README.zh-CN.md:5]: id构造器 -> Guid构造器

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/BaseRedis.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis;
+
+public class BaseRedis
+{
+    protected readonly IDistributedCacheClient DistributedCacheClient;
```
Comment (doddgu) [inline src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/BaseRedis.cs:8]: Field? Property?

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/DistributedIdGeneratorOptions.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
+// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis;
+
+public class DistributedIdGeneratorOptions : IdGeneratorOptions
+{
+    /// <summary>
+    /// When there is no available WorkerId, recover the inactive WorkerId for 2 minutes
```
Comment (doddgu) [inline src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/DistributedIdGeneratorOptions.cs:9]: recover?

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/DistributedWorkerProvider.cs [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+        await Database.SortedSetAddAsync(_inUseWorkerKey, _workerId + "", GetCurrentTimestamp());
+    }
+
+    public async Task LogOutAsync()
+    {
+        if (_workerId == null)
+            return;
+
+        var val = _workerId! + "";
```
Comment (doddgu) [inline src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/DistributedWorkerProvider.cs:99]: + ""?

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/DistributedWorkerProvider.cs [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+            return Task.FromResult(_workerId.Value);
+        }
+    }
+
+    public async Task RefreshAsync()
+    {
+        if (_workerId == null)
+            throw new MasaException("No WorkerId available");
+
+        await Database.SortedSetAddAsync(_inUseWorkerKey, _workerId + "", GetCurrentTimestamp());
```
Comment (doddgu) [inline src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/DistributedWorkerProvider.cs:91]: + ""?

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/DistributedWorkerProvider.cs [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+            _workerId,
+            DateTime.UtcNow);
+    }
+
+    private async Task<long> GetNextWorkerIdAsync()
+    {
+        var workerId = await DistributedCacheClient.HashIncrementAsync(_currentWorkerKey, 1) - 1;
+        if (workerId > _maxWorkerId)
+        {
+            var lockdb = Connection.GetDatabase(-1);
```
Comment (doddgu) [inline src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/DistributedWorkerProvider.cs:118]: -1?

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/Internal/SnowflakeExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,15 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis.Internal;
+
+internal static class SnowflakeExtensions
+{
+    public static long GetTimestamp(this DateTimeOffset dateTimeOffset, uint timeType)
+    {
+        if (timeType == 1)
```
Comment (doddgu) [inline src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/Internal/SnowflakeExtensions.cs:10]: What's the mean of the timeType?

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/ServiceCollectionExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+            if (distributedIdGeneratorOption.EnableMachineClock)
+            {
+                services.TryAddSingleton<ISnowflakeGenerator>(serviceProvider
+                    => new MachineClockIdGenerator(serviceProvider.GetRequiredService<IDistributedCacheClient>(),
+                        serviceProvider.GetRequiredService<IWorkerProvider>(),
+                        serviceProvider.GetRequiredService<IOptions<RedisConfigurationOptions>>(),
+                        distributedIdGeneratorOption));
+            }
+
+            long defaultHeartbeatinterval = 30 * 1000;
```
Comment (doddgu) [inline src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/ServiceCollectionExtensions.cs:42]: defaultHeartbeatinterval -> defaultHeartbeatInterval

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/ServiceCollectionExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+            if (distributedIdGeneratorOption.RecycleTime <= defaultHeartbeatinterval)
+            {
+                throw new ArgumentOutOfRangeException(
+                    $"{nameof(distributedIdGenerators.RecycleTime)} RecycleTime must be greater than {defaultHeartbeatinterval}");
```
Comment (doddgu) [inline src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/ServiceCollectionExtensions.cs:46]: {nameof(distributedIdGenerators.RecycleTime)} RecycleTime?

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake/BaseIdGenerator.cs [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Contrib.Data.IdGenerator.Snowflake;
+
+public abstract class BaseIdGenerator
+{
+    private readonly IWorkerProvider _workerProvider;
+
+    /// <summary>
+    /// initial timestamp
```
Comment (doddgu) [inline src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake/BaseIdGenerator.cs:11]: initial timestamp? start timestamp?

File: test/Masa.Contrib.Data.IdGenerator.Snowflake.BenchmarkDotnet.Tests/DistributedBenchmarks.cs [hunk 13]
```diff
@@ -0,0 +1,38 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the MIT License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Contrib.Data.IdGenerator.Snowflake.BenchmarkDotnet.Tests;
+
+[SimpleJob(RunStrategy.ColdStart, RuntimeMoniker.Net60, targetCount: 1000)]
```
Comment (doddgu) [inline test/Masa.Contrib.Data.IdGenerator.Snowflake.BenchmarkDotnet.Tests/DistributedBenchmarks.cs:6]: Add the benchmark result sample to the readme.

**Review by doddgu — CHANGES_REQUESTED** [review]

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/Internal/SnowflakeExtensions.cs (hunk 9 above)
Comment (zhenlei520) [reply src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/Internal/SnowflakeExtensions.cs:10]: The timestamp in Snowflake ID is milliseconds by default. By changing the timestamp type to seconds, the number of bits occupied by the timestamp can be reduced. For some scenarios with low concurrency, the second-level timestamp can be used, and the performance will be higher.

Comment (doddgu) [reply src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/Internal/SnowflakeExtensions.cs:10]: I mean you can use enum.

File: Masa.Contrib.sln (hunk 1 above)
Comment (zhenlei520) [reply Masa.Contrib.sln]: processed

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.SimpleGuid/README.zh-CN.md (hunk 3 above)
Comment (zhenlei520) [reply src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.SimpleGuid/README.zh-CN.md]: processed

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/BaseRedis.cs (hunk 4 above)
Comment (zhenlei520) [reply src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/BaseRedis.cs]: processed

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/DistributedWorkerProvider.cs (hunk 6 above)
Comment (zhenlei520) [reply src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/DistributedWorkerProvider.cs]: processed

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/DistributedIdGeneratorOptions.cs (hunk 5 above)
Comment (zhenlei520) [reply src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/DistributedIdGeneratorOptions.cs]: processed

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/DistributedWorkerProvider.cs (hunk 7 above)
Comment (zhenlei520) [reply src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/DistributedWorkerProvider.cs]: processed

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/DistributedWorkerProvider.cs (hunk 8 above)
Comment (zhenlei520) [reply src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/DistributedWorkerProvider.cs]: processed

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/ServiceCollectionExtensions.cs (hunk 10 above)
Comment (zhenlei520) [reply src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/ServiceCollectionExtensions.cs]: processed

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/ServiceCollectionExtensions.cs (hunk 11 above)
Comment (zhenlei520) [reply src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/ServiceCollectionExtensions.cs]: processed

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake/BaseIdGenerator.cs (hunk 12 above)
Comment (zhenlei520) [reply src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake/BaseIdGenerator.cs]: processed

File: test/Masa.Contrib.Data.IdGenerator.Snowflake.BenchmarkDotnet.Tests/DistributedBenchmarks.cs (hunk 13 above)
Comment (zhenlei520) [reply test/Masa.Contrib.Data.IdGenerator.Snowflake.BenchmarkDotnet.Tests/DistributedBenchmarks.cs:6]: processed

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/DistributedWorkerProvider.cs [hunk 14]
```diff
@@ -32,7 +32,7 @@ public class DistributedWorkerProvider : BaseRedis, IWorkerProvider
         ArgumentNullException.ThrowIfNull(distributedIdGeneratorOptions);
 
         _timestampType = distributedIdGeneratorOptions.TimestampType;
-        _recycleTime = distributedIdGeneratorOptions.RecycleTime;
+        _idleTimeOut = distributedIdGeneratorOptions.IdleTimeOut;
         _maxWorkerId = distributedIdGeneratorOptions.MaxWorkerId;
         _workerIdMinInterval = distributedIdGeneratorOptions.GetWorkerIdMinInterval;
         _currentWorkerKey = "snowflake.current.workerid";
```
Comment (doddgu) [inline src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/DistributedWorkerProvider.cs:38]: `Possible null reference assignment.` in 79 line.

**Review by doddgu — CHANGES_REQUESTED** [review]

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.SequentialGuid/SequentialGuidGenerator.cs (hunk 2 above)
Comment (zhenlei520) [reply src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.SequentialGuid/SequentialGuidGenerator.cs]: processed

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/DistributedWorkerProvider.cs (hunk 14 above)
Comment (zhenlei520) [reply src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/DistributedWorkerProvider.cs:38]: processed

File: src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/Internal/SnowflakeExtensions.cs (hunk 9 above)
Comment (zhenlei520) [reply src/Data/IdGenerator/Masa.Contrib.Data.IdGenerator.Snowflake.Distributed.Redis/Internal/SnowflakeExtensions.cs]: ok  processed

File: test/Masa.Contrib.Data.IdGenerator.Snowflake.Tests/IdGeneratorTest.cs [hunk 15]
```diff
@@ -1,8 +1,6 @@
 // Copyright (c) MASA Stack All rights reserved.
 // Licensed under the MIT License. See LICENSE.txt in the project root for license information.
 
-using Masa.Utils.Caching.Redis;
-
```
Comment (doddgu) [inline test/Masa.Contrib.Data.IdGenerator.Snowflake.Tests/IdGeneratorTest.cs]: `The field 'IdGeneratorTest._currentWorkerKey' is assigned but its value is never used` in line 9, 10, 11, 12.

**Review by doddgu — CHANGES_REQUESTED** [review]

**Review by doddgu — APPROVED** [review]
LGTM

**Pull request merged**